### PR TITLE
oh-my-posh: init at 8.29.4

### PR DESCRIPTION
**Pull request by ShrykeWindgrace**
###### Description of changes

Packaging `oh-my-posh` by `JanDeDobbeleer` - a prompt theme engine for any shell.

Homepage is here: https://ohmyposh.dev/
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      sha256 = "D1dR/S/qyTHyCZt7IvIhmYsq2Yh89Wm0PoGvSsko0jA=";
+    } + "/src";
+
+  vendorSha256 = "AzfNq6t+Zk4eI0v6lWTzpcEmHPxF/3Lrk9g/OgOoTmI=";
+
+  ldflags = [ "-a" "-s" "-w" "-X" "main.Version=${version}" "-extldflags" ''"-static"'' ];
+
+  tags = [ "netgo" "osusergo" "static_build" ];
+
+  dontFixup = true;
```
Comment (balsoft) [inline pkgs/tools/misc/oh-my-posh/default.nix:42]: Why? (add a comment)

**Review by balsoft — COMMENTED** [review]
I would've written the `themeInstallSteps` a bit differently, using `srcs` and `fetchzip` to get automatic unpacking and such.

File: pkgs/tools/misc/oh-my-posh/default.nix (hunk 1 above)
Comment (ShrykeWindgrace) [reply pkgs/tools/misc/oh-my-posh/default.nix:42]: @balsoft I took tags directly from [release scripts](https://github.com/JanDeDobbeleer/oh-my-posh/blob/v7.72.5/src/.goreleaser.yml); and took a guess that [definition of buildGoModule](https://github.com/NixOS/nixpkgs/blob/7beebb590d541ffa534aa34b0b163f81c3c72c2c/pkgs/development/go-modules/generic/default.nix#L13) would have a corresponding attribute.

And fixup tries to `patchELF` a statically-linked binary,  which does not have much sense, does it? I do not have a strong opinion on this. In the end of the day, the executable still works.

**Comment by ShrykeWindgrace** [comment]
> I would've written the `themeInstallSteps` a bit differently, using `srcs` and `fetchzip` to get automatic unpacking and such.

Indeed, the `srcs` and `fetchzip` look better-suited for this task. However, the files in the zip do not have their permissions set (*at all*), and `fetchzip` fails with a list of `mv: cannot open '/build/unpack/name.of.the.file.json' for reading: Permission denied` errors. I will see if I manage a workaround.

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  meta = with lib; {
+    license = licenses.mit;
+    homepage = "https://ohmyposh.dev/";
+    description = "A prompt theme engine for any shell";
+    maintainers = with maintainers; [ shrykewindgrace ];
+  };
+
+  postInstall =
+    if getThemes then ''
```
Comment (balsoft) [inline pkgs/tools/misc/oh-my-posh/default.nix:36]: `lib.optionalString`

**Comment by arjan-s** [comment]
Can't test this easily at the moment, as there is currently a merge conflict between this PR and master. Can you please fix that? Also, oh-my-posh 7.85.2 is at this moment the latest version (they put out releases very often).

**Comment by ShrykeWindgrace** [comment]
@arjan-s Thanks for the tip about the conflict. Fixed it and bumped version in the MR.

**Comment by arjan-s** [comment]
Result of `nixpkgs-review pr 170641` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>oh-my-posh</li>
  </ul>
</details>

**Review by arjan-s — APPROVED** [review]

**Comment by ShrykeWindgrace** [comment]
@arjan-s thanks for the approval!

**Comment by arjan-s** [comment]
No problem! Though I now see that tests are failing on x86_64-darwin. Could you take a look at that? According to the website, it should run on MacOS as well. Perhaps you can take a hint from [this older PR](https://github.com/NixOS/nixpkgs/pull/155639)?

**Comment by VergeDX** [comment]
close #159963.

**Comment by ShrykeWindgrace** [comment]
@arjan-s If I understand correctly, that previous PR did not work on darwin either, it had a line `platforms = platforms.linux;`. I'll try to find a workaround, but I am neither a `go` developer nor a `macOS` user.

**Comment by arjan-s** [comment]
@ShrykeWindgrace right, I missed that line! Sorry, I'm not into go or macos myself, so I'm afraid I can't help with that... :(

**Comment by x3rAx** [comment]
> If I understand correctly, that previous PR did not work on darwin either, it had a line `platforms = platforms.linux;`

This was mostly because I simply couldn't test if it works on macOS (I opened the prev. PR).

But since it's written in go, it should in theory also work on macOS aswell. Go usually creates single binary files with all dependencies inside. If there are no fancy c-bindings or other stuff involved (and I guess that would make packaging harder than just using `buildGoModule`) then it might not work on other platforms.

I don't know how the policy for the `platforms` property is. Do we set it to only the platforms we *know* it is working on or do we just set it to all except the ones we know it does not work on?

**Comment by ShrykeWindgrace** [comment]
@x3rAx First of all, please accept my apologies, I should have checked before submitting my PR, so your PR would not be in vain.

In the worst case we can grab pre-built darwin binaries of `oh-my-posh` from the GH releases, but I was told that was not the nix-way. And I do not know the `platforms` policy either.

**Comment by x3rAx** [comment]
@ShrykeWindgrace No worries, my PR had merge errors and was outdated anyways. I was busy lately and didn't find the time to fix it. I'm happy when this PR gets merged :smile_cat:

File: maintainers/maintainer-list.nix [hunk 3]
```diff
@@ -14570,4 +14570,10 @@
       fingerprint = "41EA 00B4 00F9 6970 1CB2  D3AF EF90 E3E9 8B8F 5C0B";
     }];
   };
+  shrykewindgrace = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:14573]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  modRoot = "./src";
+  vendorSha256 = "trG+w+xUlfsBkMfkM7vRxO41vXdymPFp5IYILR7hlQc=";
+
+  ldflags = [ "-a" "-s" "-w" "-X" "main.Version=${version}" "-extldflags" ''"-static"'' ];
+
+  tags = [ "netgo" "osusergo" "static_build" ];
+
+  dontFixup = true;
+
+  meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:28]: meta is always last

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    description = "A prompt theme engine for any shell";
+    maintainers = with maintainers; [ shrykewindgrace ];
+  };
+
+  postInstall = lib.optionalString getThemes ''
+    mkdir -p $out
+    cp -r ${src}/themes $out
+  '';
+}
+
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:40]: one final new line to much

File: pkgs/tools/misc/oh-my-posh/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/oh-my-posh/default.nix:26]: Please remove it. This also disabled other hooks which is most likely unwanted.

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  pname = "oh-my-posh";
+  owner = "JanDeDobbeleer";
+
+  src = fetchFromGitHub
+    {
+      inherit owner;
+      repo = pname;
+      rev = "v${version}";
+      sha256 = "OAS0Uehgy4UFkhe+C7TbFTTBJkAS3G4g8ruyS90DqOQ=";
+    };
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:18]:
```suggestion
  src = fetchFromGitHub {
    owner = "JanDeDobbeleer";
    repo = "oh-my-posh";
    rev = "v${version}";
    sha256 = "OAS0Uehgy4UFkhe+C7TbFTTBJkAS3G4g8ruyS90DqOQ=";
  };

```

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 7]
```diff
@@ -0,0 +1,40 @@
+{ lib
+, fetchFromGitHub
+, buildGoModule
+, getThemes ? false
+}:
+
+buildGoModule rec {
+  version = "7.85.2";
+  pname = "oh-my-posh";
+  owner = "JanDeDobbeleer";
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:10]:
```suggestion
  pname = "oh-my-posh";
  version = "7.85.2";
```

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 8]
```diff
@@ -0,0 +1,40 @@
+{ lib
+, fetchFromGitHub
+, buildGoModule
+, getThemes ? false
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:4]: Why don't we want the default themes?

Comment (VergeDX) [reply pkgs/tools/misc/oh-my-posh/default.nix:4]: According to docs here: https://ohmyposh.dev/docs/installation/customize  
oh-my-posh can generate specific shell scripts from json format theme,  
source the generated scripts to get theme.

That theme folder is collection listing here: https://ohmyposh.dev/docs/themes
Access by using `${generated.oh-my-posh.src}/themes/kali.omp.json` is also same.

Comment (SuperSandro2000) [reply pkgs/tools/misc/oh-my-posh/default.nix:4]: Those themes are in total not bigger than 8MB. We can safely copy them to the output all the time and remove that flag.

**Comment by ShrykeWindgrace** [comment]
@SuperSandro2000 , @VergeDX  thank you for your feedback, I appreciate it. I am currently on vacation with no access to my nix setup, I will be able to address you suggestions next Wednesday at the earliest.

However, the major blocking point still stands - current derivation does not build on x86_64-darwin, and I do not the expertise to fix it.

**Comment by SuperSandro2000** [comment]
> current derivation does not build on x86_64-darwin, and I do not have the expertise to fix it.

You can mark it broken on that platform. Not a big problem.

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  meta = with lib; {
+    description = "A prompt theme engine for any shell";
+    homepage = "https://ohmyposh.dev/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ shrykewindgrace ];
+    # it should on darwin, too, but I do not have the expertise to make it work on nix
+    platforms = platforms.linux;
+  };
+
+}
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:37]:
```suggestion
  };
}
```

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  modRoot = "./src";
+  vendorSha256 = "iBVehyfJ2WX/nSBbXeOD0ozjLSVn5LKDNgbYVopIr4w=";
+
+  ldflags = [ "-a" "-s" "-w" "-X" "main.Version=${version}" "-extldflags" ''"-static"'' ];
+
+  tags = [ "netgo" "osusergo" "static_build" ];
+
+  postInstall = ''
+    mkdir -p $out
+    cp -r ${src}/themes $out
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:25]:
```suggestion
    cp -r ${src}/themes $out/share/oh-my-posh/
```

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    cp -r ${src}/themes $out/share/oh-my-posh/
+  '';
+
+  meta = with lib; {
+    description = "A prompt theme engine for any shell";
+    homepage = "https://ohmyposh.dev/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ shrykewindgrace ];
+    # it should on darwin, too, but I do not have the expertise to make it work on nix
+    platforms = platforms.linux;
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:35]:
```suggestion
    platforms = platforms.darwin;
```
Maybe add a short comment about the error.

Also please squash all package commits into ``oh-my-posh: init at 8.1.0``.

Comment (SuperSandro2000) [reply pkgs/tools/misc/oh-my-posh/default.nix:35]: Yeah, I meant broken.

**Comment by arjan-s** [comment]
Hi @ShrykeWindgrace are you still working on this?

**Comment by ShrykeWindgrace** [comment]
@arjan-s Hi, thanks for the ping. Nothing of the essence changed - I can bump version, 8.11.*, if I recall correctly, but current derivation still does not compile on darwin and I still have neither idea nor means to fix it=)

I marked the derivation as broken on darwin, codestyle seems ok to me, so, in my opinion, the ball is on the side of people with merge rights=)

**Comment by arjan-s** [comment]
Cool, thanks. I'll be patient for the merge then :)

File: pkgs/tools/misc/oh-my-posh/default.nix [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    description = "A prompt theme engine for any shell";
+    homepage = "https://ohmyposh.dev/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ shrykewindgrace ];
+    # it should work on darwin, too, but I do not have the expertise to make it work on nix
+    # it has something to do with -lpthread
+    broken = stdenv.isDarwin;
+  };
+
+}
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/oh-my-posh/default.nix:39]:
```suggestion
  };
}
```

Comment (ShrykeWindgrace) [reply pkgs/tools/misc/oh-my-posh/default.nix]: Added in the last iteration

File: maintainers/maintainer-list.nix (hunk 3 above)
Comment (SuperSandro2000) [reply maintainers/maintainer-list.nix:14931]: This still must be sorted.

**Comment by ShrykeWindgrace** [comment]
Closing since https://github.com/NixOS/nixpkgs/pull/184670 was since merged.
